hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -3128,6 +3137,44 @@ func (r *secretJanitor) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return ctrl.Result{}, cleanup(ctx, r.Client, secret)
 }
 
+// shouldKeepOldUserData determines if the old user data should be kept.
+// For AWS < 4.16, we keep the old userdata Secret so old Machines during rolled out can be deleted.
+// Otherwise, deletion fails because of https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3805.
+// TODO (alberto): Drop this check when support for old versions without the fix is not needed anymore.
+func (r *NodePoolReconciler) shouldKeepOldUserData(ctx context.Context, hc *hyperv1.HostedCluster) (bool, error) {
+	if hc.Spec.Platform.Type != hyperv1.AWSPlatform {
+		return false, nil
+	}
+
+	// If there's a curren version in status, be conservative and assume that one is the one running CAPA.
+	releaseImage := hc.Spec.Release.Image
+	if hc.Status.Version != nil {
+		if len(hc.Status.Version.History) > 0 {
+			releaseImage = hc.Status.Version.History[0].Image
+		}
+	}
+
+	pullSecretBytes, err := r.getPullSecretBytes(ctx, hc)
+	if err != nil {
+		return true, fmt.Errorf("failed to get pull secret bytes: %w", err)
+	}
+
+	releaseInfo, err := r.ReleaseProvider.Lookup(ctx, releaseImage, pullSecretBytes)
+	if err != nil {
+		return true, fmt.Errorf("failed to lookup release image: %w", err)
+	}
+	hostedClusterVersion, err := semver.Parse(releaseInfo.Version())
+	if err != nil {
+		return true, err
+	}
+
+	if hostedClusterVersion.LT(semver.MustParse("4.16.0")) {
```

Comment:
this will also exclude all the 4.16 nightlies, but I guess that it is expected/wanted

Reply:
this is saying less than 4.16

Reply:
yep. Just saying that this includes any 4.16 nightly, if it makes any difference. But we probably don't care. 
https://go.dev/play/p/YNZ4-qiy7Zt

Reply:
I see, yes that's desired as far as I can tell